### PR TITLE
Add tests for SubgroupDispatch

### PR DESCRIPTION
**Pull request by dekimir**
`SubgroupDispatch` is a new feature in v1.1 adding a new capability, new execution models, and new instructions.

Extended the testing framework to allow testing with different SPIR-V versions.

@antiagainst @AWoloszyn @dneto0 @Qining PTAL.

### CODE REVIEW

File: test/OpcodeRequiresCapabilities.cpp [hunk 1]
```diff
@@ -135,6 +135,8 @@ INSTANTIATE_TEST_CASE_P(
         ExpectedOpCodeCapabilities{SpvOpEmitStreamVertex,
                                    mask(SpvCapabilityGeometryStreams)},
         ExpectedOpCodeCapabilities{SpvOpTypeNamedBarrier,
-                                   mask(SpvCapabilityNamedBarrier)}), );
+                                   mask(SpvCapabilityNamedBarrier)},
+        ExpectedOpCodeCapabilities{SpvOpGetKernelMaxNumSubgroups,
```
Comment (dneto0) [inline test/OpcodeRequiresCapabilities.cpp]: Also need OpGetKernelLocalSizeForSubgroupCount

File: test/OperandCapabilities.cpp [hunk 2]
```diff
@@ -37,12 +37,19 @@ struct EnumCapabilityCase {
   uint64_t expected_mask;
 };
 
-using EnumCapabilityTest = ::testing::TestWithParam<EnumCapabilityCase>;
+// Test fixture for testing EnumCapabilityCases.
+template <spv_target_env env>
+class EnumCapabilityTest : public ::testing::TestWithParam<EnumCapabilityCase> {
+ protected:
+  spv_target_env env_ = env;  // Target environment to use in tests.
```
Comment (dneto0) [inline test/OperandCapabilities.cpp]: Should be const.  

Could also be static but I don't care about that part.

File: test/TextToBinary.ModeSetting.cpp [hunk 3]
```diff
@@ -140,22 +140,30 @@ TEST_F(OpEntryPointTest, WrongModel) {
...
 // Test OpExecutionMode
 
-using OpExecutionModeTest = spvtest::TextToBinaryTestBase<
-    ::testing::TestWithParam<EnumCase<SpvExecutionMode>>>;
+template <spv_target_env env>
+class OpExecutionModeTest
+    : public spvtest::TextToBinaryTestBase<
+          ::testing::TestWithParam<EnumCase<SpvExecutionMode>>> {
+ protected:
+  spv_target_env env_ = env;
```
Comment (dneto0) [inline test/TextToBinary.ModeSetting.cpp]: Should be const

File: test/TextToBinary.ModeSetting.cpp [hunk 4]
```diff
@@ -190,15 +198,26 @@ INSTANTIATE_TEST_CASE_P(
...
         {CASE(ContractionOff), {}},
-    }),);
+    }), );
+
+INSTANTIATE_TEST_CASE_P(
+    TextToBinaryExecutionMode, OpExecutionModeTestV11,
+    ::testing::ValuesIn(std::vector<EnumCase<SpvExecutionMode>>{
+        {CASE(SubgroupSize), {12}},
+        {CASE(SubgroupsPerWorkgroup), {64}},
+        {CASE(LocalSize), {64, 1, 2}},
```
Comment (dneto0) [inline test/TextToBinary.ModeSetting.cpp]: Might be worth saying that in this test, LocalSize and below are spot checking features available in 1.0.
Above that line are features new to 1.1

File: test/TextToBinary.SubgroupDispatch.cpp [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+using ::testing::Eq;
+using ::testing::Matcher;
+using ::testing::Property;
+using ::testing::SizeIs;
+using std::vector;
+
+// Creates a matcher of SPIR-V word vectors that matches a single instruction
+// with the given opcode and length. TODO(dekimir): move this into TestFixture
+// and DRY other tests that perform this match in longhand.
+Matcher<vector<uint32_t>> IsOpcode(SpvOp opcode, uint16_t length) {
```
Comment (dneto0) [inline test/TextToBinary.SubgroupDispatch.cpp]: It's more than an opcode check.  It's a check for the opcode word and the instruction length.  So at least the name is wrong.

I dislike the number of layers of abstraction that you have to plough through to know what the test is testing.  That's high complexity for low payoff.
We already have many tests that check the well-formedness of the instructions when the initial word (opcode,length) is specified.  So I think you gain very little over just explicitly matching against an explicitly constructed vector.

Please don't DRY the rest of the tests in this way.

**Comment by dneto0** [comment]
-1
- Missing OpGetKernelLocalSizeForSubgroupCount in OpcodeRequiresCapabilities.cpp
- Two places where const should be used
- Please avoid increases in complexity in test code that have little to no benefit.

File: test/TextToBinary.SubgroupDispatch.cpp (hunk 5 above)
Comment (dekimir) [inline test/TextToBinary.SubgroupDispatch.cpp]: The number of layers is actually quite low, and the benefit substantial: this lets us check a `CompiledInstructions` result (with meaningful failure messages) in the shortest, most readable, least error-prone fashion.  Other tests either have multiple lines to assert the vector length and the first element (assuming they even test both), or they rely on `ElementsAre(spvOpcodeMake(...), _, _, _, _)`, which gets cumbersome with 7-argument instructions like the ones tested here.

I'll change the name to something like `IsSingleInstruction`.

File: test/OpcodeRequiresCapabilities.cpp (hunk 1 above)
Comment (dekimir) [inline test/OpcodeRequiresCapabilities.cpp]: See comment on line 114.

**Comment by dekimir** [comment]
Addressed feedback so far.

File: test/TextToBinary.SubgroupDispatch.cpp [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+namespace {
+
+using ::testing::AllOf;
+using ::testing::Eq;
+using ::testing::Matcher;
+using ::testing::Property;
+using ::testing::SizeIs;
+using std::vector;
+
+// Creates a matcher of SPIR-V word vectors that matches a single instruction
```
Comment (dneto0) [inline test/TextToBinary.SubgroupDispatch.cpp]: No reason to make this new abstraction. Most of the other tests in the suite use a matcher like:
```
  Eq(MakeInstruction(the_opcode, { the, list, of, operands }))
```
which is fully explicit and easier to read.  There are about 250 such examples.

**Comment by dneto0** [comment]
-1.  Please follow the predominant pattern in the rest of the tests.

File: test/TextToBinary.SubgroupDispatch.cpp (hunk 6 above)
Comment (dekimir) [inline test/TextToBinary.SubgroupDispatch.cpp]: K, will leverage `MakeInstruction()` instead.

Comment (dekimir) [inline test/TextToBinary.SubgroupDispatch.cpp]: Done.

**Comment by dneto0** [comment]
+2

**Comment by dekimir** [comment]
Squashed and merged to master via 2ea54f5.

**Pull request closed without merging**